### ISSUE TITLE
 Support for futures and async calls 

### ISSUE THREAD

**Issue by surajprak** [mlua-rs/rlua#113]
It will a desirable feature to make call to rust function in non blocking manner and work seamlessly with futures. Current implementation does not yield if i make future call in rust from lua stack. Therefore, it makes futures not usable. Can you please provide Lua yield support with futures?

**Comment by kyren** [comment]
Which are you asking for, being able to yield in a `Future` from Lua, or being able to yield in Lua from a `Future`?  Could you sketch out what you want to do in a little more detail?

Lua coroutines can be pretty gnarly, and you're going to run into a lot of things that are not possible currently with `rlua` (anything involving lua_callk / lua_pcallk / lua_yieldk), but generally there is *a* way to make an API that plays nicely with Lua coroutines.

If you're asking for some kind of deep integration with Rust generators, I think that might need to wait until generators are stabilized, and I'm not exactly sure what something like this would look like.

**Comment by kyren** [comment]
Just so you don’t get the wrong idea, I don’t think interacting with async specifically is something I want to add in rlua.  I can see making rlua more seamless with whatever stable generators look like, but I don’t think anything should end up in rlua that is aware of e.g. Future specifically.  It seems very out of scope, I’d rather build things that just make things like this easy to do rather than offer some kind of opinionated one size solution.  That should be the job of an associated crate if anything.

For the question about how to restart yielded Lua coroutines when the future they’re waiting on is ready, that’s a really good question.  I imagine they would have to yield a value back to RustA that identified what they are waiting on so that RustA can resume them at the appropriate time.  Maybe a userdata with methods is easier to deal with than a callback for the RustB portion?

But it’s good that you pointed that out, because I think those are the sorts of decisions that should *not* be decided by rlua, but I can help you work through what that solution would look like.

**Comment by surajpayu** [comment]
I agree to you that this library should be generic. I am trying to work on this as per your suggestion.
I have couple of questions?
1)Is there an option to create a new lua thread(coroutine) from rust? I can see in example that coroutine is created in lua script? It would be good if a wrapper over lua_newthread would be available in rust. I want to create coroutine inside my rust code and store its handle. when my future called in RustB is ready, i wanted to use this coroutine handle in rust to resume it. If it can be done, there is no need to create an loop in lua. Simply when luaA calls RustB, it will yield (from lua) but will be resumed by rust?
2) Also, is it possible to directly load a lua script using rlua, instead of embedding everything in rust code?

**Comment by kyren** [comment]
> 1)Is there an option to create a new lua thread(coroutine) from rust? I can see in example that coroutine is created in lua script?

[Context::create_thread](
https://docs.rs/rlua/0.16.2/rlua/struct.Context.html#method.create_thread)

> Also, is it possible to directly load a lua script using rlua, instead of embedding everything in rust code?

You can load a file and point `rlua` at a `&[u8]` which is the file's contents.  There ought to be a way to pass `rlua` a `Read` impl so that it doesn't require loading the entire file's contents into memory, but I haven't written it yet.

**Comment by surajpayu** [comment]
Thanks for quick response. Also, is there some macro function available to implement FromLua and ToLua for my struct. May be some examples on how to do it, will be helpful.

**Comment by kyren** [comment]
> Thanks for quick response. Also, is there some macro function available to implement FromLua and ToLua for my struct? May be some examples on how to do it, will be helpful.

Not yet, but I plan on doing this soon.  I think there's an external crate that might provide one?

**Comment by surajpayu** [comment]
which one?

**Comment by kyren** [comment]
Looks like none that are up to date, but I meant rlua-table-derive and rlua_serde, I have not used either though.

**Comment by surajpayu** [comment]
Also, how do i make working my struct seamless. Lets take below example
struct Transaction{
     id: String,
}

impl Transaction{
 pub fn parse(data: String)->Result<Self, String>{
        let t: Result<Transaction,  serde_json::error::Error> = serde_json::from_str(&data[..]);
        match t{
            Ok(t_inner) => Ok(t_inner),
            Err(err) => {
                error!("Error while converting input {} to transaction", err);
                Err("Input is not parsable to a transaction object".to_string())
            }
        }
    }
}


is it possible that if i am working with Transaction(a struct in rust) object in lua. I would be able to call its functions. For example, in lua
t:parse("213")

**Comment by kyren** [comment]
You mean like a userdata with methods?  https://github.com/kyren/rlua/blob/e351bf57fa4c3985109d6674a8b6c35f9890da16/examples/guided_tour.rs#L166

**Comment by surajpayu** [comment]
So rust structs are mapped to userdata in lua? Is my understanding correct? 
Also, is it also possible to use static method of rust struct in lua.
For e.g
```
Transaction::total_count() 
```

**Comment by kyren** [comment]
The `UserData` trait is used for userdata values that are given to Lua, static methods like that wouldn't have a userdata value associated with them.  You could make a table that has regular functions on it for that kind of API.

**Comment by surajpayu** [comment]
@kyren thanks for the suggestion. 

Is rlua supposed to work with lua 5.3 only? I was more comfortable in working with lua 5.2

**Comment by kyren** [comment]
> Is rlua supposed to work with lua 5.3 only?

It only works with Lua 5.3 right now, this will probably change after Lua 5.4 is released so that `rlua` can work with either Lua 5.3 or 5.4.

I'm not sure exactly *how* I'm going to do this, because it would be pretty ridiculous to include Lua 5.3 *AND* Lua 5.4 sources inside the crate.  Either I will have to only bundle Lua 5.4 and just ensure that the API usage is 5.3 compatible, or if I can't make the API used both 5.3 and 5.4 compatible there will have to be some kind of build switches.

If I go down this road, I could also check what it would take to be compatible with Lua 5.2 and MAYBE even 5.1 / LuaJIT.  Probably this is a ways off and a lot of work.

**Comment by leo60228** [comment]
I have a full example of this at https://github.com/leo60228/rlua-async. I was thinking about packaging it into a crate, but I'm not sure how the API should work...

**Comment by khvzak** [comment]
You could try [mlua](https://github.com/khvzak/mlua) for this. This is rlua 0.15 fork and I recently added async/await support.
